pkg/dns/treecache.go

```diff
@@ -0,0 +1,312 @@
+				if subpath == "*" || subpath == "any" {
+					nextNodesToExplore = append(nextNodesToExplore, node)
+				} else {
+					if val, ok := node.Entries[subpath]; ok {
+						retval = append(retval, val)
+					} else {
+						childNode := node.ChildNodes[subpath]
+						if childNode != nil {
+							nextNodesToExplore = append(nextNodesToExplore, childNode)
+						}
+					}
+				}
+			}
+			nodesToExplore = nextNodesToExplore
+			break
+		}
+
+		if subpath == "*" || subpath == "any" {
+			for _, node := range nodesToExplore {
+				for subkey, subnode := range node.ChildNodes {
+					if !strings.HasPrefix(subkey, "_") {
+						nextNodesToExplore = append(nextNodesToExplore, subnode)
+					}
+				}
+			}
+		} else {
+			for _, node := range nodesToExplore {
+				childNode := node.ChildNodes[subpath]
+				if childNode != nil {
+					nextNodesToExplore = append(nextNodesToExplore, childNode)
+				}
+			}
+		}
+		nodesToExplore = nextNodesToExplore
+	}
+
+	for _, node := range nodesToExplore {
+		for _, val := range node.Entries {
+			retval = append(retval, val)
+		}
+	}
+
+	return retval
+}
+
+func (cache *TreeCache) GetEntries(recursive bool, path ...string) []interface{} {
+	cache.m.RLock()
+	defer cache.m.RUnlock()
+	childNode := cache.GetSubCache(path...)
+	if childNode == nil {
+		return nil
+	}
+
+	retval := [][]interface{}{{}}
+	childNode.appendValues(recursive, retval)
+	return retval[0]
+}
+
+func (cache *TreeCache) DeletePath(path ...string) bool {
+	if len(path) == 0 {
+		return false
+	}
+	cache.m.Lock()
+	defer cache.m.Unlock()
+	if parentNode := cache.GetSubCache(path[:len(path)-1]...); parentNode != nil {
+		if _, ok := parentNode.ChildNodes[path[len(path)-1]]; ok {
+			delete(parentNode.ChildNodes, path[len(path)-1])
+			return true
+		}
+	}
+	return false
+}
+
+func (tn *TreeCache) DeleteEntry(key string, path ...string) bool {
+	tn.m.Lock()
+	defer tn.m.Unlock()
+	childNode := tn.GetSubCache(path...)
+	if childNode == nil {
+		return false
+	}
+	if _, ok := childNode.Entries[key]; ok {
+		delete(childNode.Entries, key)
+		return true
+	}
+	return false
+}
+
+func (tn *TreeCache) appendValues(recursive bool, ref [][]interface{}) {
+	for _, value := range tn.Entries {
+		ref[0] = append(ref[0], value)
+	}
+	if recursive {
+		for _, node := range tn.ChildNodes {
+			node.appendValues(recursive, ref)
+		}
+	}
+}
+
+func (tn *TreeCache) ensureChildNode(path ...string) *TreeCache {
+	childNode := tn
+	for _, subpath := range path {
+		newNode := childNode.ChildNodes[subpath]
+		if newNode == nil {
+			newNode = NewTreeCache()
+			childNode.ChildNodes[subpath] = newNode
+		}
+		childNode = newNode
+	}
+	return childNode
+}
+
+func ensureDir(path string, perm os.FileMode) error {
+	s, err := os.Stat(path)
+	if err != nil || !s.IsDir() {
+		return os.Mkdir(path, perm)
+	}
+	return nil
+}
+
+func getMD5(b []byte) []byte {
+	h := md5.New()
+	h.Write(b)
+	return []byte(fmt.Sprintf("%x", h.Sum(nil)))
+}
+
+func main() {
```

Comment:
This should be moved as Test* functions to a test file? The code here looks like a test. Also, you expect this code to be used as a library right? Not binary?